src/components/common/animations/ReorderAnimation.tsx

```diff
@@ -0,0 +1,120 @@
+import { useTransition } from '@react-spring/web';
+
+import { useState } from 'react';
+import { animated } from 'react-spring';
+
+const ANIMATION_DURATION_MS = 450;
+
+type DataElementType = { elementType: number };
+
+export type TransitionData<T extends DataElementType> = {
+  key: string;
+  height?: number;
+  marginBottom: number;
+  y?: number;
+  data: T;
+};
+
+type Props<T extends DataElementType> = {
+  isAnimating: boolean;
+  elements: TransitionData<T>[];
+  renderElement: (
+    item: TransitionData<T>,
+    onHeightChange: (key: string, height: number) => void
+  ) => JSX.Element;
+};
+
+export const ReorderAnimation = <T extends DataElementType>({
+  isAnimating,
+  elements,
+  renderElement,
+}: Props<T>) => {
+  const [totalHeight, setTotalHeight] = useState<number>(0);
+  const elementHeights = new Map<string, number>();
+  // this variable is used to set the next element to the good y
+  let transitionHeight = 0;
+  const transitionAnimation = isAnimating
+    ? {
+        from: { height: 0, opacity: 0 },
+        leave: { height: 0, opacity: 0 },
+        enter: ({ y, height }: { y: number; height: number }) => ({
+          y,
+          height,
+          opacity: 1,
+        }),
+        update: ({ y, height }: { y: number; height: number }) => ({
+          y,
+          height,
+        }),
+      }
+    : {
+        from: (_item: TransitionData<T>, _: number) => ({
+          opacity: 0,
+        }),
+        enter: { opacity: 1 },
+      };
+
+  const transitions = useTransition(
+    elements.map((e) => {
+      transitionHeight += e.marginBottom;
+      return {
+        ...e,
+        y: transitionHeight - e.marginBottom,
+      };
+    }),
+    {
+      // the key must be unique and not change to keep the same component for the animation.
+      key: (item: TransitionData<T>) => `${item.key}`,
+      ...transitionAnimation,
+      config: { duration: ANIMATION_DURATION_MS },
+    }
+  );
+
+  const handleElementHeight = (key: string, height: number) => {
+    elementHeights.set(key, height);
```

Comment:
This `elementHeights` is only used inside this funcition it seems ?

Reply:
Yes, it is only used in this method. The map is set outside, because totalHeight must be recomputed each time an element's height changed.

Using a map allows to "track" the height of each element and to recompute the total height directly.

If you have another solution to do that I'm interested 😄